### PR TITLE
feat(frequencies): add support for QuasiDistribution type

### PR DESCRIPTION
**Pull request by akotil**
### Summary

Refactor frequencies utilities to generalize to `QuasiDistribution`s and add further functionality to convert from `Counts` to `QuasiDistribution`s.

### CODE REVIEW

File: pr_toolbox/quantum/results/frequencies.py
Comment (pedrorrivero) [inline pr_toolbox/quantum/results/frequencies.py:23]: Great job!! 💯

File: pr_toolbox/quantum/results/frequencies.py
Comment (pedrorrivero) [inline pr_toolbox/quantum/results/frequencies.py]: LOVE THIS! 

Check this [reference](https://docs.python.org/3/library/functools.html#functools.singledispatch), this [PEP](https://peps.python.org/pep-0443/), or this [tutorial](https://www.geeksforgeeks.org/function-overloading-with-singledispatch-functools/) for a native python way of doing _single dispatch_ (possibly more performant) 😉 I'll let you decide what you consider the best approach to be!

File: pr_toolbox/quantum/results/frequencies.py
Comment (pedrorrivero) [inline pr_toolbox/quantum/results/frequencies.py]: Maybe it is good to avoid the word _'mapping'_ in the error message since functions building on top do not use it (e.g. `bitflip`). 
```suggestion
    raise TypeError(f"Invalid frequencies type. Expected `Counts` or `QuasiDistribution` but got {type(frequencies)} instead.")
```

File: pr_toolbox/quantum/results/frequencies.py
Comment (pedrorrivero) [inline pr_toolbox/quantum/results/frequencies.py]: Very good! But this should be a separate PR 😅  Two reasons:
1. It is an independent contribution doing its own thing, so splitting it up in two PRs will make this more apparent and ensure that it shows up as an entire new entry in the change log.
2. Will help keep this PR smaller.

Also, would you consider naming it `counts_to_quasi_dists` instead? I try to use verbs for functions and nouns for variables, however in this case it seems unnecessary to use the word _'convert'_ as it is implied —mostly interested in hearing your thoughts, feel free to fight me back if you disagree. 

Otherwise should be ready to merge if you port it; test looks good as well. I will merge regardless of what you decide to do with the name 🙃

File: pr_toolbox/quantum/results/frequencies.py
Comment (pedrorrivero) [inline pr_toolbox/quantum/results/frequencies.py]: This line is copied from the above code but the variable name not updated:
```suggestion
    frequencies_dict: dict[int, float] = defaultdict(lambda: 0)
```

File: pr_toolbox/quantum/results/frequencies.py
Comment (pedrorrivero) [inline pr_toolbox/quantum/results/frequencies.py]: I wonder if it would make sense having a function that does this for a `dict` type directly. This would spare us some code duplicity.

Single dispatch could then call itself recursively, meaning that if it receives `Counts` it will convert them to a frequencies dict (i.e. `counts.int_outcomes()` and call itself again with this new argument type. The result will then be wrapped in a `Counts` object and returned. Something similar would happen for `QuasiDistribution`.

File: test/unit/quantum/results/test_frequencies.py
Comment (pedrorrivero) [inline test/unit/quantum/results/test_frequencies.py:79]: Missing test cases for `QuasiDistribution` here as well as in `TestMaskCounts`

File: test/unit/quantum/results/test_frequencies.py
Comment (pedrorrivero) [inline test/unit/quantum/results/test_frequencies.py]: You also need to check that `shots` and `std_dev` are correctly set for `QuasiDistribution`

File: test/unit/quantum/results/test_frequencies.py
Comment (pedrorrivero) [inline test/unit/quantum/results/test_frequencies.py]: See how easy it is to have comprehensive type tests 🙃 Glad that the ping-pong occurred early on! 🏓

**Review by pedrorrivero — CHANGES_REQUESTED** [review]
This is looking really good! Just some minor comments and a cool way of doing things that you may actually enjoy learning about 🙃 

Thanks a lot!

File: test/unit/quantum/results/test_frequencies.py
Comment (akotil) [reply test/unit/quantum/results/test_frequencies.py]: Indeed! 🙂

File: pr_toolbox/quantum/results/frequencies.py
Comment (akotil) [reply pr_toolbox/quantum/results/frequencies.py]: Done! You can find the dedicated PR in #3 .

File: pr_toolbox/quantum/results/frequencies.py
Comment (akotil) [reply pr_toolbox/quantum/results/frequencies.py]: Love the `singledispatch` decorator! Thanks a lot for introducing me to this!

File: pr_toolbox/quantum/results/frequencies.py
Comment (pedrorrivero) [inline pr_toolbox/quantum/results/frequencies.py]: There is only one distribution, so it should be singular (the word distribution already encompasses several frequencies):
```suggestion
def map_quasi_dist(quasi_dist: QuasiDistribution, mapper: Callable) -> QuasiDistribution:
```

File: test/unit/quantum/results/test_frequencies.py
Comment (pedrorrivero) [inline test/unit/quantum/results/test_frequencies.py]: If you are only going to do this, would it be cleaner to pass the right type from the beginning instead of passing along the type as well? 🤔

**Review by pedrorrivero — CHANGES_REQUESTED** [review]
I think a cleaner structure, without any explicit type checks, would be:
```python
@singledispatch
def map_frequencies(frequencies, mapper):
    """Default behavior."""
    raise TypeError(...)

@map_frequencies.register
def map_dict(frequencies: dict, mapper) -> dict:
    """Operates on dict input, returns dict."""
    # TODO: check valid dict.keys() and dict.values()
    ...  # Apply map

@map_frequencies.register
def map_counts(counts: Counts, mapper) -> Counts:
    """Operates on Counts input, returns Counts."""
    frequencies: dict = counts.int_outcomes()
    frequencies = map_frequencies(frequencies, mapper)
    return Counts(frequencies)

@map_frequencies.register
def map_quasi_dist(quasi_dist: QuasiDistribution, mapper) -> QuasiDistribution:
    """Operates on QuasiDistribution input, returns QuasiDistribution."""
    frequencies: dict = dict(quasi_dist)
    frequencies = map_frequencies(frequencies, mapper)
    return QuasiDistribution(
        frequencies, 
        shots=quasi_dist.shots,
        stddev_upper_bound=quasi_dist.stddev_upper_bound
    )
```

You'll need to complete the type annotations, docstrings, missing logic... 

You can leave the `TODO` as a comment if you don't have time to implement it.

File: test/unit/quantum/results/test_frequencies.py
Comment (akotil) [reply test/unit/quantum/results/test_frequencies.py]: It has the purpose of preventing the initialization on both the inputs and outputs in the parameter set and keeping the code a bit more shallow. But one could also infer the type from the frequencies input, that is in fact cleaner!

File: pr_toolbox/quantum/results/frequencies.py
Comment (pedrorrivero) [inline pr_toolbox/quantum/results/frequencies.py]:
```suggestion
################################################################################
## FREQUENCIES
################################################################################
@singledispatch
```

File: pr_toolbox/quantum/results/frequencies.py
Comment (pedrorrivero) [inline pr_toolbox/quantum/results/frequencies.py]:
```suggestion
def map_frequencies(
    frequencies: FrequenciesLike | dict, mapper: Callable
) -> FrequenciesLike | dict:
```

File: pr_toolbox/quantum/results/frequencies.py
Comment (pedrorrivero) [inline pr_toolbox/quantum/results/frequencies.py]:
```suggestion
```

File: pr_toolbox/quantum/results/frequencies.py
Comment (pedrorrivero) [inline pr_toolbox/quantum/results/frequencies.py]:
```suggestion
```

File: pr_toolbox/quantum/results/frequencies.py
Comment (pedrorrivero) [inline pr_toolbox/quantum/results/frequencies.py]:
```suggestion
def map_counts(counts: Counts, mapper: Callable) -> Counts:
```

File: pr_toolbox/quantum/results/frequencies.py
Comment (pedrorrivero) [inline pr_toolbox/quantum/results/frequencies.py:15]:
```suggestion
from collections import defaultdict
from collections.abc import Callable
```

File: pr_toolbox/quantum/results/frequencies.py
Comment (pedrorrivero) [inline pr_toolbox/quantum/results/frequencies.py]:
```suggestion
def map_quasi_dist(quasi_dist: QuasiDistribution, mapper: Callable) -> QuasiDistribution:
```

**Review by pedrorrivero — DISMISSED** [review]
Thanks a lot, this is awesome work @akotil!! 🥇 

I'll take it from here

**Review by pedrorrivero — APPROVED** [review]

**Pull request merged**
